src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,617 @@
+        # AttributePath = [[Endpoint = Specific Endpoint, Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value to the DUT
+
+        self.print_step(1, "Send Request Message to read one attribute on a given cluster and endpoint")
+
+        # endpoint = [(0, Clusters.Objects.Descriptor.Attributes.ServerList)]
+        read_request_1 = await self.default_controller.ReadAttribute(self.dut_node_id, server_list_attr_path)
+        returned_endpoints = read_request_1[0].keys()
+        # returned_endpoints = dict_keys([<class 'chip.clusters.Objects.Descriptor'>])
+        # Check if chip.clusters.Objects.Descriptor is in output
+        asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
+        # Check if ServerList is in nested output
+        asserts.assert_in(server_list_attr, read_request_1[0][descriptor_obj], "ServerList not in output")
+
+        # Step 2
+        # TH sends the Read Request Message to the DUT to read all attributes on a given cluster and Endpoint
+        # AttributePath = [[Endpoint = Specific Endpoint, Cluster = Specific ClusterID]]
+        # On receipt of this message, DUT should send a report data action with the attribute value to the DUT.
+
+        self.print_step(2, "Send Request Message to read all attributes on a given cluster and endpoint")
+        # endpoint = [(0, Clusters.Objects.Descriptor)]
+        read_request_2 = await self.default_controller.ReadAttribute(self.dut_node_id, descriptor_obj_path)
+        returned_endpoints = read_request_2[0].keys()
+        # returned_endpoints = dict_keys([<class 'chip.clusters.Objects.Descriptor'>])
+        # Check if chip.clusters.Objects.Descriptor is in output
+        asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
+
+        # Step 3
+
+        # TH sends the Read Request Message to the DUT to read an attribute from a cluster at all Endpoints
+        # AttributePath = [[Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the Endpoints to the DUT.
+        self.print_step(3, "Send Request Message to read one attribute on a given cluster at all endpoints")
+        # endpoint = [Clusters.Objects.Descriptor.Attributes.ServerList]
+        read_request_3 = await self.default_controller.ReadAttribute(self.dut_node_id, [server_list_attr])
+        for i in range(3):
+            returned_endpoints = read_request_3[i].keys()
+
+            # returned_endpoints = dict_keys([<class 'chip.clusters.Objects.Descriptor'>])
+            # Check if chip.clusters.Objects.Descriptor is in output
+            asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
```

Comment:
add checks that ONLY requested attribute and cluster are returned, but that the endpoint list matches.
you can do that easily as a set comparison
(code not guaranteed to be exactly correct since i'm coding in a text box, but something like this...)
asserts.assert_equal(set(returned_endpoints.keys(), set(<whatever_cluster_you_requested>), 'some msg')

same for attribute - check that there's just the one and it matches.

same for endpoints - check that the read_request keys matches the expected endpoint list exactly. With the current code, if you get an extra endpoint returned you won't catch it.